File: omrchecker-js/packages/core/src/processors/layout/field/bubbleField.ts
/**
 * Bubble field classes.
 *
 * TypeScript port of src/processors/layout/field/bubble_field.py
 * Defines BubbleField and BubblesScanBox for bubble-based OMR fields.
 */

import { ZERO_MARGINS } from '../../../utils/constants';
import { defaultDump } from '../../../utils/parsing';
import { FieldBlock } from '../fieldBlock/base';
import { Field, ScanBox } from './base';
import { BubbleFieldDrawing, type FieldDrawing } from './fieldDrawing';

/**
 * Scan box for individual bubbles.
 *
 * Container for a Point Box on the OMR.
 * field_label is the point's property - field to which this point belongs to.
 * It can be used as a roll number column as well. (eg roll1)
 * It can also correspond to a single digit of integer type Q (eg q5d1).
 */
export class BubblesScanBox extends ScanBox {
  public bubbleFieldType: string;
  public bubbleDimensions: [number, number];
  public bubbleValue: string;

  constructor(
    fieldIndex: number,
    field: BubbleField,
    origin: [number, number],
    bubbleValue: string
  ) {
    const dimensions = field.bubbleDimensions;
    const margins = ZERO_MARGINS;
    super(fieldIndex, field, origin, dimensions, margins);

    this.bubbleFieldType = field.bubbleFieldType;
    this.bubbleDimensions = field.bubbleDimensions;
    this.bubbleValue = bubbleValue;
    this.name = `${this.fieldLabel}_${this.bubbleValue}`;
  }

  /**
   * Serialize to JSON.
   */
  toJSON(): Record<string, unknown> {
    return {
      field_label: this.fieldLabel,
      bubble_value: this.bubbleValue,
      name: this.name,
      x: this.x,
      y: this.y,
      origin: defaultDump(this.origin),
    };
  }
}

/**
 * Bubble field class.
 *
 * Extends Field for bubble-based fields (MCQ, integer, etc.).
 */
export class BubbleField extends Field {
  public bubbleDimensions: [number, number];
  public bubbleValues: string[];
  public bubblesGap: number;
  public bubbleFieldType: string;

  constructor(
    direction: 'horizontal' | 'vertical',
    emptyValue: string,
    fieldBlock: FieldBlock,
    fieldDetectionType: 'BUBBLES_THRESHOLD',
    fieldLabel: string,
    origin: [number, number]
  ) {
    // This ensures parent class is fully initialized before accessing self properties
    super(direction, emptyValue, fieldBlock, fieldDetectionType, fieldLabel, origin);

    if (
      !fieldBlock.bubbleDimensions ||
      !fieldBlock.bubbleValues ||
      fieldBlock.bubblesGap === undefined ||
      !fieldBlock.bubbleFieldType
    ) {
      throw new Error('FieldBlock missing bubble properties');
    }

    // Set subclass-specific properties after super()
    this.bubbleDimensions = fieldBlock.bubbleDimensions;
    this.bubbleValues = fieldBlock.bubbleValues;
    this.bubblesGap = fieldBlock.bubblesGap;
    this.bubbleFieldType = fieldBlock.bubbleFieldType;
  }

  /**
   * Setup scan boxes for this bubble field.
   * Creates a BubblesScanBox for each bubble value.
   *
   * Uses fieldBlock parameters instead of self properties to allow proper initialization order.
   * This method is called from the base constructor before subclass properties are set.
   */
  setupScanBoxes(fieldBlock: FieldBlock): void {
    // Use fieldBlock parameters instead of self properties
    // This allows setupScanBoxes to work even if called before subclass properties are set
    const bubbleValues = fieldBlock.bubbleValues;
    const bubbleDimensions = fieldBlock.bubbleDimensions;
    const bubblesGap = fieldBlock.bubblesGap;
    const bubbleFieldType = fieldBlock.bubbleFieldType;

    if (!bubbleValues) {
      throw new Error('bubbleValues is required and must not be empty');
    }

    // populate the field bubbles
    const h = this.direction === 'vertical' ? 1 : 0;

    const bubblePoint: [number, number] = [...this.origin];
    this.scanBoxes = [] as BubblesScanBox[];

    // Temporarily set properties so BubblesScanBox constructor can access them.
    // setupScanBoxes() is called from base Field constructor (before BubbleField sets these).
    // BubblesScanBox constructor accesses field.bubbleDimensions and field.bubbleFieldType.
    // These will be set again (to the same values) after super() returns in BubbleField constructor.
    // Using non-null assertions since constructor validates these.
    this.bubbleDimensions = bubbleDimensions!;
    this.bubbleFieldType = bubbleFieldType!;

    for (let fieldIndex = 0; fieldIndex < bubbleValues.length; fieldIndex++) {
      const bubbleValue = bubbleValues[fieldIndex];
      const bubbleOrigin: [number, number] = [...bubblePoint];
      const scanBox = new BubblesScanBox(fieldIndex, this, bubbleOrigin, bubbleValue);
      this.scanBoxes.push(scanBox);
      bubblePoint[h] += bubblesGap!;
    }

    // Note: We don't restore original values because they'll be overwritten in the constructor anyway.
  }

  /**
   * Get drawing instance for bubble fields.
   */
  getDrawingInstance(): FieldDrawing {
    return new BubbleFieldDrawing(this);
  }
}

File: src/processors/layout/field/bubble_field.py
from src.processors.layout.field.base import Field, ScanBox
from src.processors.layout.field.field_drawing import BubbleFieldDrawing
from src.utils.constants import ZERO_MARGINS
from src.utils.parsing import default_dump


class BubbleField(Field):
    def __init__(
        # ruff: noqa: PLR0913
        self,
        direction,
        empty_value,
        field_block,
        field_detection_type,
        field_label,
        origin,
    ) -> None:
        # Call super().__init__() first (proper initialization order)
        super().__init__(
            direction,
            empty_value,
            field_block,
            field_detection_type,
            field_label,
            origin,
        )

        # Set subclass-specific properties after super().__init__()
        # This ensures proper initialization order and avoids accessing self before parent init
        self.bubble_dimensions = field_block.bubble_dimensions
        self.bubble_values = field_block.bubble_values
        self.bubbles_gap = field_block.bubbles_gap
        self.bubble_field_type = field_block.bubble_field_type
        # self.plot_bin_name = field_label

    def get_drawing_instance(self):
        return BubbleFieldDrawing(self)

    def setup_scan_boxes(self, field_block) -> None:
        # Use field_block parameters instead of self properties
        # This allows setup_scan_boxes to work even if called before subclass properties are set
        bubble_values = field_block.bubble_values
        bubble_dimensions = field_block.bubble_dimensions
        bubbles_gap = field_block.bubbles_gap
        bubble_field_type = field_block.bubble_field_type

        if not bubble_values:
            raise ValueError("bubble_values is required and must not be empty")

        # populate the field bubbles
        h = 1 if (self.direction == "vertical") else 0

        bubble_point = self.origin.copy()
        self.scan_boxes: list[BubblesScanBox] = []

        # Temporarily set properties so BubblesScanBox constructor can access them.
        # setup_scan_boxes() is called from base Field constructor (before BubbleField sets these).
        # BubblesScanBox constructor accesses field.bubble_dimensions and field.bubble_field_type.
        # These will be set again (to the same values) after super() returns in BubbleField constructor.
        self.bubble_dimensions = bubble_dimensions
        self.bubble_field_type = bubble_field_type

        for field_index, bubble_value in enumerate(bubble_values):
            bubble_origin = bubble_point.copy()
            scan_box = BubblesScanBox(field_index, self, bubble_origin, bubble_value)
            self.scan_boxes.append(scan_box)
            bubble_point[h] += bubbles_gap

        # Note: We don't restore original values because they'll be overwritten in the constructor anyway.


class BubblesScanBox(ScanBox):
    """(TODO: update docs).

    Container for a Point Box on the OMR
    field_label is the point's property- field to which this point belongs to
    It can be used as a roll number column as well. (eg roll1)
    It can also correspond to a single digit of integer type Q (eg q5d1).
    """

    def __init__(self, field_index, field: BubbleField, origin, bubble_value) -> None:
        dimensions = field.bubble_dimensions
        margins = ZERO_MARGINS
        super().__init__(field_index, field, origin, dimensions, margins)
        self.bubble_field_type = field.bubble_field_type
        self.bubble_dimensions = field.bubble_dimensions

        self.shifts = [0, 0]
        self.bubble_value = bubble_value
        self.name = f"{self.field_label}_{self.bubble_value}"
        self.plot_bin_name = self.field_label

    # Make the class serializable
    def to_json(self):
        return {
            key: default_dump(getattr(self, key))
            for key in [
                "field_label",
                "bubble_value",
                # for item_reference_name
                "name",
                "x",
                "y",
                "origin",
                # "plot_bin_name",
                # "bubble_field_type",
                # "field_index",
                # "bubble_dimensions",
            ]
        }
